(cap:op_matematicas)=
# Operações matemáticas e tipos

Em sua maneira mais simples, Python pode ser usado como uma calculadora que segue as mesmas convenções de ordem de operações com que está acostumado.

```{note}
Seção da documentação oficial relevante para este capítulo: https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex
```

## Operações matemáticas básicas

Aqui estão as operações binárias (operam em dois números) e unárias (operam em um número) básicas em Python. Preste atenção na presença do ponto decimal nas entradas e nas respostas.

**Soma: `+`**

In [1]:
2 + 4

6

In [2]:
3.7 + 4.9

8.600000000000001

In [3]:
3.7 + 5

8.7

**Subtração: `-`**

Quando este símbolo é utilizado sozinho com um número, indica que é negativo (-5)

In [4]:
2 - 4

-2

In [5]:
3.7 - 4.9

-1.2000000000000002

In [6]:
3.7 - 4

-0.2999999999999998

**Multiplicação: `*`**

In [7]:
2 * 4

8

In [8]:
3.7 * 4.9

18.130000000000003

In [9]:
3.7 * 5

18.5

**Divisão: `/`**

In [10]:
4 / 2

2.0

In [11]:
3.7 / 4.9

0.7551020408163265

In [12]:
3.7 / 5

0.74

**Divisão inteira: `//`**

In [13]:
4 // 2

2

In [14]:
3.7 // 4.9

0.0

In [15]:
3.7 // 2

1.0

**Potenciação: `**`**

In [16]:
2 ** 4

16

In [17]:
3.7 ** 4.9

608.3989388909348

In [18]:
4 ** 0.5

2.0

**Operação módulo: `%`**

Retorna o resto de uma divisão inteira. Por exemplo

10 // 4 = 2 * 4 + 2, então 10 % 4 = 2

In [19]:
2 % 4

2

In [20]:
4 % 2

0

In [21]:
10 % 4

2

In [22]:
3.7 % 2

1.7000000000000002

Quando o símbolo `-` é utilizado sozinho perto de um número (operação unária), indica que é um número negativo. Existe o equivalente para números positivos, `+`, mas não é muito utilizado.

In [23]:
-4 - -7

3

In [24]:
-4 - +7

-11

In [25]:
3 ** (-1)

0.3333333333333333

In [26]:
3 ** -1

0.3333333333333333

O operador módulo se comporta um pouco diferente de outras linguagens quando se trata de números negativos. Leia o conteúdo [aqui](https://stackoverflow.com/questions/3883004/how-does-the-modulo-operator-work-on-negative-numbers-in-python) para entender melhor, se desejar. Em suma, o resultado é sempre positivo se o número à direita for positivo.

In [27]:
-10 % 4

2

In [28]:
10 % -4

-2

## Comparações e booleanos

Para fazer comparações entre dois números, podemos utilizar operadores binários comumente utilizados em matemática. O resultado de uma comparação é um valor `booleano`, que nada mais é que um valor verdadeiro `True` ou falso `False`.

**Menor que `<`**

In [29]:
10 < 2

False

In [30]:
2 < 10

True

In [31]:
2 < 2

False

**Menor ou igual que `<=`**

In [32]:
10 <= 2

False

In [33]:
2 <= 10

True

In [34]:
2 <= 2

True

**Maior que `>`**

In [35]:
10 > 2

True

In [36]:
2 > 10

False

In [37]:
2 > 2

False

**Maior ou igual que `>=`**

In [38]:
10 >= 2

True

In [39]:
2 >= 10

False

In [40]:
2 >= 2

True

**Diferente de `!=`**

Em várias linguagens, inclusive Python, o símbolo `!` é utilizado como "negação", invertendo o significado do que vem depois dele. Por isso está presente neste símbolo.

In [41]:
10 != 2

True

In [42]:
2 != 10

True

In [43]:
2 != 2

False

**Igual a `==`**

Muito importante, note que o símbolo de igual é `==`, com o sinal `=` repetido! Isso o diferencia do operador de designação (assignment), que será abordado [no próximo capítulo](cap:vars).

In [44]:
10 == 2

False

In [45]:
2 == 10

False

In [46]:
2 == 2

True

## Ordem de operações

De forma simplificada, a ordem de operações matemáticas está a seguir, da maior para menor precedência:

1. Operações dentro de parênteses.
2. Potenciação. Porém, quando se refere imediatamente a um número com o operador negativo, este toma precedência. Vide exemplo acima, `3 ** -1` é igual a `3 ** (-1)`
3. Operador negativo e positivo.
4. Multiplicação, divisão, divisão inteira, módulo
5. Adição e subtração
6. Deslocamentos de bits (explicados em uma [seção extra](extra:bitwise))
7. Operadores bit-a-bit (explicados em uma [seção extra](extra:bitwise))
8. Comparações

Mais informações podem ser encontradas [na documentação oficial aqui](https://docs.python.org/3/reference/expressions.html#operator-precedence)

Existe mais um operador que não foi abordado aqui, e será abordado no futuro. É a multiplicação matricial, realizado com `@`, p.e. `mat1 @ mat2`.

```{note}
Parênteses podem, e devem, ser adicionados para realçar e clarear o significado de certas expressões. Em Python, operações entre parênteses podem ter qualquer formatação interna, permitindo que você divida visualmente um cálculo em várias linhas, se assim desejar.
```

Aqui estão alguns exemplos do uso de parênteses

In [47]:
-4 - (-2)

-2

In [48]:
-4 - (-2) ** 2

-8

In [49]:
(-4 - (-2)) ** 2

4

In [50]:
27 ** (1/3)

3.0

In [51]:
(25/5) * 9 + 32 > (20/5) * 9 + 32

True

In [52]:
((25/5) * 9 + 32 > (20/5)) * 9 + 32

41

In [53]:
((25/5) * 9 + 32 < (20/5)) * 9 + 32

32

## Dois tipos numéricos `int` e `float` e o tipo booleano

Note que nos resultados dos anteriores, ocasionalmente ocorre o aparecimento de pontos ou não nos valores de entrada e nos resultados. Quando um número não possui um ponto de separação decimal, é um número inteiro `int` e quando tem o ponto, é um `float`, um número de ponto flutuante. No cotidiano, não classificamos os números dessa maneira. Se nos perguntam "quanto é 10 dividido por 4", dizemos "2,5" e não nos importamos com o fato de termos passado do domínio dos inteiros $\mathbb Z$ para os reais $\mathbb R$. Porém, dentro do computador, esses números são representados de uma maneira muito distinta. Quando algumas operações são feitas com eles, o Python realiza uma conversão quando necessário, para seguir com nosso entendimento intuitivo.

No caso da divisão (internamente chamada de `true division`, o resultado é sempre um `float`. Porém, em certas situações[^0], queremos uma divisão somente da parte inteira, logo o operador `//` existe.

Existem três maneiras de representar números em Python básico, os números inteiros `int` (1, 5, 10, 1000, -500)[^1] e os números de ponto flutuante `float` (1.1, 1000.3, 1E-7, etc.)[^2] e os números complexos, que serão abordados em breve. Para criar um `int`, basta não colocar um ponto no número, e para criar um `float`, basta colocar um (e só um) ponto.

Aqui, `int` e `float` são os *tipos* desses números, que por sua vez são *objetos*. Um *tipo* é uma descrição, um conjunto de regras que ditam como interpretar um trecho de memória e como interagir com outros tipos, e um *objeto* como a concretização dessa descrição. Em breve, notaremos que se não há nenhuma regra sobre como dois tipos devem interagir, aparecerá uma mensagem de erro.

```{note}
Em Python, tudo é um objeto, e todo objeto possui um tipo.
```

Como um exemplo simplista, imagine uma bancada cheia de abacates. Eles todos diferem, mas você sabe que são todos abacates, pois possui um modelo mental do que um abacate é (p.e. casca verde-escura, polpa verde-amarelada gordurosa com sabor adocicado e uma semente grande no centro). Aqui, o *conceito* de abacate é seu *tipo*, e as frutas em si são *objetos*.

Um computador temos que representar todo tipo de dado com zeros e uns (bits), a diferenciação de blocos de memória (sequências de bits) é feita somente pelo tipo associado ao bloco. Um mesmo bloco pode significar números, texto, imagens, música, etc., dependendo do contexto. Por exemplo, a sequência de bits '1100001' pode significar tanto o número 97 quando o caracter 'a'. Já o `float` "97." é representado pela seguinte sequência, gigante, pois o tamanho dos `float`s em Python é 64 bits: `0100000001011000010000000000000000000000000000000000000000000000`.

O tipo booleano só pode ter dois valores, como já mencionado. `True` ou `False`. Note que nos dois últimos exemplos, uma operação estranha foi realizada. Colocando um par de parênteses, a ordem das operações foi alterada e a comparação entre os parênteses foi avaliada antes das demais operações de multiplicação e soma. Mas mesmo assim um resultado numérico foi obtido. Isso é porque `True`, em contextos numéricos, é equivalente a `1` e `False` é equivalente a `0`.[^3] Em uma [seção futura](sec:truthy_falsy), irei abordar a direção oposta, i.e., quando certos tipos podem ser interpretados como verdadeiro ou falso.

[^0]: Por exemplo, se você quer saber quantas laranjas cabem em uma caixa, utilizar a divisão `/` não irá retornar um valor satisfatório. Não cabem 6,3 laranjas, somente 6.

[^1]: Em Python, `int`s são representados como sequências de bits sem um limite teórico, somente prático (memória de seu computador). Isso é diferente de outras línguas que possuem um limite de, por exemplo, 32 ou 64 bits nos números inteiros, possibilitando representar $2^{32}$ ou $2^{64}$ valores diferentes.

[^2]: A representação e operações com `float`s estão descritas na norma [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754).

[^3]: Por favor, não utilize isso no lugar de 0 ou 1.

## Maneiras alternativas de representar números

Podemos adicionar *underlines* a números inteiros para facilitar nosso entendimento. Por exemplo, o 1 milhão pode ser representado como

In [54]:
1_000_000

1000000

E a posição dos *underlines* não precisa ser regular, mas não podem ser consecutivos. É relativamente comum, em softwares financeiros, utilizar números inteiros da menor unidade monetária ao invés de `float`s, por razões de precisão. Logo, R$123,45 é representado como 12345 centavos, ou

In [55]:
123_45

12345

`float`s também podem receber *underlines* para facilitar o entendimento. Diferente de `int`s, `float`s podem ser escritos em notação científica abreviada, da seguinte maneira:

In [56]:
2.34E-5 * 9.23E4 / 10_000E-2

0.021598199999999998

Vimos que para criar um número inteiro, somente precisamos escrevê-lo como estamos habituados, na base 10. Existem outras maneiras de representar números inteiros em bases diferentes. Em Python, conseguimos representá-los nas bases binária (2), octal (8) e hexadecimal (16). Para isso, adicionamos um prefixo. Para representar o número 14 (base 10), podemos fazê-lo das seguintes maneiras:

* Binário: `0b1110`, pois $1 \times 2^3 + 1 \times 2^2 + 1 \times 2^1 + 0 \times 2^0 = 14$
* Octal: `0o16`, pois $1 \times 8^1 + 6 \times 8^0 = 14$
* Hexadecimal: `0xe`, pois $14 \times 16^0 = 14$[^4]

[^4]: Na base hexadecimal, A=10, B=11, C=12, D=13, E=14, F=15. As letras podem ser maiúsculas ou não.

Para comprovar:

In [57]:
0b1110

14

In [58]:
0o16

14

In [59]:
0xe

14

É importante entender que tal representação é para *nosso benefício*, não do computador. Para o computador, tudo é armazenado da maneira binária.

(sec:limit_float)=
## Limitações de números de ponto flutuante

Note que em vários exemplos mostrados acima, algumas operações pareciam fornecer valores errados. Por exemplo, `3.7 * 4.9` deveria fornecer `18.13` mas fornece `18.130000000000003`. Isso ocorre porque a representação interna de `float`s utiliza base 2. Qualquer fração que utiliza como base um número diferente de 2 é um decimal que se repete (como 3 é para 10). Isso resulta num "bug" clássico, que possui até um [site dedicado a ele](https://0.30000000000000004.com/). Honestamente, isso não é extremamente relevante para o uso comum deste curso. Mas se você for utilizar `float`s em sistemas financeiros ou simulações numéricas, é bom saber que essa limitação existe. Em Python, o pacote [`decimal`](https://docs.python.org/3/library/decimal.html?highlight=decimal#module-decimal) fornece uma maneira mais precisa de lidar com números de ponto flutuante escritos em base decimal.

In [60]:
0.1 + 0.2

0.30000000000000004

Outra limitação de `float`s, especialmente crítica no uso de ferramentas computacionais para problemas científicos, é a dificuldade de utilizar o operador de igualdade `==`. Veja os exemplos a seguir.

In [61]:
0.1 + 0.2 == 0.3

False

In [62]:
0.1 + 0.2 == 0.30000000000000004

True

In [63]:
0.2 + 0.2 == 0.4

True

In [64]:
1E-20 == 0.0

False

Vemos novamente que `0.1 + 0.2 != 0.3`, mas se você utiliza a representação completa de `0.1 + 0.2`, o resultado é verdadeiro. Porém, se você utiliza números representáveis na base binária, como `0.2`, o problema anterior não se manifesta. E por último, veja que um número muito pequeno, diminuto, que em muitos contextos seria considerado 0, não é, de fato, zero. Limite-se a utilizar `==` ou `!=` com `int`s somente.

Logo, é recomendado que você utilize um valor de tolerância, seja relativo ou absoluto, para fazer comparações com `float`s. Por exemplo, se adotarmos $1\times10^{-7}$ como um valor de tolerância absoluta para a diferença entre dois números, podemos testar essa condição com:

In [65]:
(1.0032E-5 - 1.0E-5) < 1E-7

True

In [66]:
(1.32E-5 - 1.0E-5) < 1E-7

False

A tolerância relativa, por outro lado, vê o percentual de diferença entre os dois números. Adotando 0,1%, temos que

In [67]:
(1.0032E-5 - 1.0E-5) / 1.0E-5 < 0.001

False

In [68]:
(1.0032E-5 - 1.0E-5) / 1.0E-5 > 0.001

True

A escolha do tipo de tolerância irá depender do problema em questão. Veja que, neste exemplo, a mesma operação "passou" no teste de tolerância absoluta mas falhou no teste de tolerância relativa. Como tais testes não são comumente utilizados em conjunto, essa "discrepância" não é um problema. Os valores utilizados aqui são relativamente comuns em vários pacotes científicos.

## Números complexos `complex`

Para criar um número complexo, basta colocar o sufixo `j` logo depois de um `int` ou `float`. Certas operações com `int`s e `float`s podem produzir números complexos também. Todas as operações descritas anteriormente funcionam também com números complexos, exceto a divisão inteira `//`, módulo `%` e operações de comparação exceto igualdade `==` e diferença `!=`. Se você tentar aplicar essa operação, um `TypeError` será lançado. Erros são a maneira que Python tem de informar que alguma coisa inesperada aconteceu. Neste caso, como há a definição dessas operações no tipo `complex`, então o erro lançado é um `TypeError`.

In [69]:
1j

1j

In [70]:
7j * ((6 + 2j) / 2j + (3 - 2j)**2)

(105+42j)

In [71]:
1.23 + 1j

(1.23+1j)

In [72]:
(-1)**(1/2)

(6.123233995736766e-17+1j)

In [73]:
1 // 1j

TypeError: unsupported operand type(s) for //: 'int' and 'complex'

In [74]:
1 % 1j

TypeError: unsupported operand type(s) for %: 'int' and 'complex'

In [75]:
1j % 1j

TypeError: unsupported operand type(s) for %: 'complex' and 'complex'

In [76]:
1j < 2j

TypeError: '<' not supported between instances of 'complex' and 'complex'

## Exercícios resolvidos

### Calculando frações molares

Considere a seguinte solução:

1. 10g de glicerina 80%, massa molar 92,094 g/mol. Considere o resto da massa como água destilada.
2. 0,1g de [lauril glicosídeo](https://en.wikipedia.org/wiki/Lauryl_glucoside), massa molar 348,48 g/mol.
3. 5g de água destilada, massa molar 18,015 g/mol.

Calcule a fração molar de lauril glicosídeo dada pela equação

$$
   x_i = \frac{n_i}{\sum_j^N n_j} 
$$

onde $x_i$ é a fração molar da espécie $i$, $n_i$ é o número de mols da espécie $i$ e $N$ é o número total de espécies.

Para resolver esse problema, devemos primeiro calcular o número de mols de cada espécie e depois substituir os valores na fórmula. O número de mols é calculado pela massa dividida pela massa molar. O detalhe é que a glicerina contém 20% de água, então dos 10 g de glicerina adicionada, 8 são de glicerina e 2 são de água. Somando o número de mols de todas as espécies, utilizamos isso para dividor o número de mols do lauril glicosídeo e obtemos a resposta.

In [77]:
(
    (0.1 / 348.48) / 
    ((5 + 2) / 18.015 + 0.1 / 348.48 + 8 / 92.094)
)

0.000603213294407632

Note que eu propositalmente adicionei parênteses extras e espaços entre os operadores e números para auxiliar na leitura da equação. Compare com a seguinte solução, idêntica no resultado, mas sem os espaços e parênteses. Qual você acha que é mais legível?

In [78]:
(0.1/348.48)/((5+2)/18.015+0.1/348.48+8/92.094)

0.000603213294407632

### Determine se um número é múltiplo de outro

Determine se os seguintes números são divisíveis por 7

* 7940
* 2009
* 8419
* 3328
* 8545
* 6639
* 9618
* 9718
* 3761
* 3556

Para resolver isso, vamos utilizar o operador módulo `%`. Lembre-se que ele retorna o resto da divisão inteira e, se um número é múltiplo de outro, não há resto na divisão, ou seja, `a % b == 0`.

In [79]:
(
    7940 % 7,
    2009 % 7,
    8419 % 7,
    3328 % 7,
    8545 % 7,
    6639 % 7,
    9618 % 7,
    9718 % 7,
    3761 % 7,
    3556 % 7
)

(2, 0, 5, 3, 5, 3, 0, 2, 2, 0)

Vemos então que *2009*, *9618* e *3556* são divisíveis por 7, pois tem resto igual a zero. Eu coloquei todos os cálculos em uma célula só por conveniência. O resultado da célula é tecnicamente uma tupla (`tuple`), que será abordada num [capítulo futuro](sec:tuple). Por enquanto, entenda como uma sequência de números separados, com uma vírgula como separador.

Você pode se perguntar: mas por que saber se um número é múltiplo de outro é interessante? O operador módulo opera num "ciclo", vai de 0 até *n-1*. Então, se você quiser que algo se repita a cada *n* vezes, pode utilizar `% n` e checar se é igual a zero. Eu já utilizei isso para decorar um gráfico a cada 10 pontos, por exemplo.

Outra utilidade do operador módulo é determinar se um número é par ou ímpar. Isso é igual a determinar se um número é divisível por 2.

(sec:prob_gas_ideal)=
### Fórmula do gás ideal

A fórmula do gás ideal é 

$$pV = nRT$$

considerando 3,4g de hélio a 77°F em um frasco de 35mL, calcule a pressão em bar. Considere a constante dos gases igual a 8,314472 J/(K mol). Considere a massa molar de hélio como 4 g/mol.

Rearranjando e isolando $p$, temos:

$$
p = \frac{nRT}{V}
$$

Convertendo as unidades para SI, temos que:

$$
1 mL = 1 \cdot 10^{-3} L = 1 \cdot 10^{-3} \cdot (dm)^3 = 1 \cdot 10^{-3} \cdot \left(10^{-1}\right)^3 \cdot m^3 = 10^{-6} m^3
$$

$$
1 bar = 100000 Pa
$$

$$
(C/5) = (F-32)/9
$$

$$
K = C + 273.15
$$

In [80]:
(
    ((3.4 / 4) * 8.314472 * ( (77 - 32) / 9 * 5 + 273.15 ) /
    (35 * 1E-6))
    / 1E5
)

602.0331007942857

### Quantos tomates cabem numa caixa?

Considere tomates de 3,14 cm de diâmetro, esféricos, e desejamos colocá-los em 1 camada dentro de uma caixa retangular de lado 30 cm por comprimento 40 cm (internos). Quantos tomates cabem na caixa?

Adotando uma maneira ingênua de resolver o problema, podemos pensar que é só dividir o lado e comprimento pelo diâmetrodo tomate, depois multiplicar esses 2 valores.

In [81]:
30 / 3.14 * 40 / 3.14

121.70879143170107

Arredondando para baixo, resulta em 121 tomates cabendo na caixa. Porém, note que temos que arredondar antes, ao calcular quantos tomates cabem nas duas dimensões.

In [82]:
30 // 3.14

9.0

In [83]:
40 // 3.14

12.0

In [84]:
9*12

108

Veja a diferença na figura. Os tomates em vermelho escuro não cabem na caixa. Se quiser confirmar o resultado, pode contar o número de tomates vermelho claro (ou abrir o arquivo .svg no Inkscape e selecioná-los).

![tomates](./Imagens/tomates.png)

## Exercícios extra

Estes exercícios requerem conteúdo abordado em classes futuras.

### Extendendo uma classe

```{note}
Este exercício requer conhecimento de **classes**, abordado [aqui](cap:oop).
```

Vimos que números complexos não implementam os operadores `%` e `//`. Extenda a classe `complex` com essa funcionalidade, mesmo que não faça sentido matemático.

In [ ]:
%load "./Soluções dos exercícios/complex_mod.py"

## Fórmula do gás ideal com o pacote `pint`

Resolva o problema do gás ideal mostrado anteriormente, mas utilizando o pacote `pint` de conversão de unidades.

In [ ]:
%load "./Soluções dos exercícios/pint.py"